Widescreen HUD Hack Patch for Echoes

Nystrata opened the pull request:
Leaving this as a draft for now because this will depend on other PRs for other repositories and there's likely a lot to be tested and change

- open-prime-rando https://github.com/randovania/open-prime-rando/pull/108 (You are here!) 
- ppc-asm https://github.com/henriquegemignani/ppc-asm/pull/72
- Randovania https://github.com/randovania/randovania/pull/7458

## What this does: 
* Via a dol patch, Render the game in 16:9 (ported from [gamemasterplc's ntsc 16:9 gecko code](https://wiki.dolphin-emu.org/index.php?title=Metroid_Prime_2:_Echoes_(GC)#NA)) 
* Via provided [edited FRMEs from my widescreenhud mod](https://github.com/Nystrata/EchoesWidescreenHUD), adjust the HUD/Visor widgets to better match the 16:9 aspect ratio
* For FRME replacement, there is a new schema entry for Widescreen Patch  
   For DOL patches, new entry in EchoesDolPatchesData
* This works for both NTSC-U and PAL versions of the game. (I ported gmplc's code to PAL)

## To-Do
- Done! ~~Open PR for the other repos~~
- This still requires a *lot* of testing. I found empty space in memory, but I don't have a good way to make sure that it doesn't cause issues.


![Screenshot 2024-07-27 150003](https://github.com/user-attachments/assets/5cb6ea1c-08a9-4166-af46-f3db44cd66e7)

Nystrata committed:
WidescreenHUD Patch

intentional comet

Nystrata committed:
Renamed FRME files to match names in PWE

Nystrata committed:
Summary (Required)

Description

Nystrata committed:
Update dol_patches.py

Nystrata committed:
Update dol_patches.py

Nystrata committed:
Summary (required)

Description

Nystrata committed:
Summary (required)

Description

Nystrata committed:
Merge branch 'main' into WidescreenHUDPatch

Nystrata committed:
Seperate out apply_widescreen_hack_patch into it's own function

Makes it so RDV can just call it directly instead of relying on a config file

Nystrata committed:
Remove prints from widescreen_hud\__init__

Nystrata committed:
Clean up

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Nystrata committed:
Make configurable via EchoesDolPatchesData

Nystrata committed:
Merge branch 'WidescreenHUDPatch' of https://github.com/Nystrata/open-prime-rando into WidescreenHUDPatch

Nystrata committed:
Add docstrings, pass version.description instead of the entire version object

Nystrata committed:
Restore vanila instructions when widescreen hack is disabled

Restore vanilla behavior when widescreen hack is disabled
Needed until RDV is updated to no longer cache patched DOL files

Nystrata committed:
Merge branch 'randovania:main' into WidescreenHUDPatch

Nystrata committed:
Merge branch 'main' into WidescreenHUDPatch

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Nystrata commented:
I think, at this stage, it's ready to go as is

Future plans do involve stubbing out DrawWidescreen: 
* It'll get rid of the black bars in *some* cutscenes (thought, it doesn't get all of them, so I have to research how to hide the blackbars in other cutscenes)
* It'll give me space in the DOL to store the extra PPC, instead of finding random and possibly unreliable unused empty space.

However, I'm holding off of this until the patcher can get a fresh vanilla dol each time, since I don't think there's a good way to restore vanilla behavior when the options is turned off in RDV.

Nystrata committed:
Merge branch 'randovania:main' into WidescreenHUDPatch

Nystrata committed:
Merge branch 'randovania:main' into WidescreenHUDPatch

duncathan commented on src/open_prime_rando/dol_patching/echoes/dol_patches.py:
```diff
@@ -500,3 +500,50 @@ def apply_map_door_changes(door_symbols: MapDoorTypeAddresses, dol_file: DolFile
+            culling_replacement = 0x8030256C
+            culling_insertion = 0x80418E8C
+            viewport_replacement = 0x8036D684
+            viewport_insertion = 0x80003748
```
these should be new fields added to `EchoesDolVersion` rather than a custom implementation like this. see other functions in this file for reference

Nystrata replied:
Moved these addresses to dol_versions.py

duncathan commented on src/open_prime_rando/echoes/widescreen_hud/__init__.py:
```diff
@@ -0,0 +1,30 @@
+
+def apply_widescreen_hud(editor: PatcherEditor):
+    """
+    Replaces certain FRMEf files to adjust HUD/Visor widgets to better fit a 16:9 screen aspect ratio
```
>FRMEf

typo?

Nystrata replied:
Typo. Fixed

duncathan commented on src/open_prime_rando/echoes/widescreen_hud/__init__.py:
```diff
@@ -0,0 +1,30 @@
+    elif editor.does_asset_exists(0xB5CF0C19) and editor.does_asset_exists(0xD9D58FA5):
+        detectedversion = "pal"
+    else:
+        raise NotImplementedError
```
again, don't love this. we already have some amount of version handling for dol patches and I'd like to see that leveraged unless there's a very specific reason it needs to be handled this way (in which case please add a comment explaining that reason)

Nystrata replied:
https://github.com/randovania/open-prime-rando/pull/108/commits/a7be495d7c3b8e86c227f7d2672c79e740d4832d

Add comment explaining manual version detection 
```py
 # TODO: Properly do version detection
 # Currently detecting game version by existence of version-specific assets here, since PAK patching doesn't seem
 # to have version detection,unlike DOL patching that does
 ```

Nystrata replied:
https://github.com/randovania/open-prime-rando/pull/108/commits/6dfe9451f1d4c24519573b872698989fc88ce5f7
Changed to no longer need the notion of versioning - if the asset exists in game, replace it. If it doesn't ignore it.

I checked to make sure that the custom asset IDs are different if the files are different, so there's no problematic overlap between the versions. ([My Notes](https://github.com/Nystrata/EchoesWidescreenHUD/wiki/Notes#gguisyspak))

Nystrata committed:
Update __init__.py - Fixed comment typo

Fixed comment typo

Nystrata committed:
Move addresses to dol_versions.py

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Nystrata committed:
Fixed variable typos

widecreen -> widescreen

Nystrata committed:
Add comment explaining manual version detection

Nystrata committed:
Comment typo fix

Nystrata committed:
Check if asset exists, instead of needing to deal with game versions

Nystrata committed:
Merge branch 'randovania:main' into WidescreenHUDPatch

henriquegemignani commented on src/open_prime_rando/echoes/widescreen_hud/__init__.py:
```diff
@@ -0,0 +1,18 @@
+    Replaces certain FRME files to adjust HUD/Visor widgets to better fit a 16:9 screen aspect ratio
+    """
+    for asset_id, filename in WIDESCREEN_HUD_ASSETS.items():
+        widescreen_assets = custom_asset_path().joinpath("widescreen_hud")
```
this line could be outside the for loop

henriquegemignani commented on src/open_prime_rando/dol_patching/echoes/dol_patches.py:
```diff
@@ -500,3 +509,40 @@ def apply_map_door_changes(door_symbols: MapDoorTypeAddresses, dol_file: DolFile
+        dol_file.write_instructions(
+            viewport_insertion,
+            [
+                lis(r14, 0x3FAA),
```
Can you document a bit excessively what these instructions are supposed to do, and what you are patching?

Think of having to come back to this code in 3 years.

Nystrata committed:
Merge branch 'randovania:main' into WidescreenHUDPatch

Nystrata committed:
Update __init__.py

Nystrata committed:
Rename variables to be more accurate, added comments

Nystrata committed:
Replace culling section with a simple float value change, updated comments

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Nystrata committed:
Remove usage of storing to r2

Nystrata commented:
Setting this as a draft again - reworking a lot of the dol patches and there might be things incomplete

Nystrata committed:
Rework culling code, rename variables for clarity

Nystrata commented:
Should be good to go now, opening for review

henriquegemignani commented on src/open_prime_rando/dol_patching/echoes/dol_versions.py:
```diff
@@ -85,6 +86,13 @@
         double_damage_vfx=0x80017F28,
+        widescreen_render=WidescreenRenderAddresses(
+            culling_original_instruction_address=0x8030256C,
+            culling_custom_frustum_value_offset=-0x3490,  # assume r2 = 0x804223c0
```
r2 is already tracked via the `sda2_base` variable, meaning you could have these two offsets be just addresses and do the math in runtime.

Also remove the suffix `_address` from all attributes because everything is an address.

Nystrata committed:
Merge branch 'randovania:main' into WidescreenHUDPatch

Nystrata committed:
Make all entries in WidescreenRenderAddresses class absolute addresses

Addresses review https://github.com/randovania/open-prime-rando/pull/108/files/bda18b6b8fbc3ede346a2e92e29e2b87f4ab04eb

Nystrata committed:
Fix arithmetic typos

duncathan reviewed (changes requested):
looks mostly good, just a couple documentation nitpicks

duncathan commented:
ah, looks like this still needs a new ppc-asm release @henriquegemignani

Nystrata commented on src/open_prime_rando/echoes/schema.json:
```diff
@@ -73,6 +73,11 @@
         },
+        "widescreen_hud": {
+            "type": "boolean",
+            "description": "Enable 16:9 widescreen HUD patch",
```
Without the new FRMEs, the HUD/Visor elements remain in 4:3, like so:  
I don't think it's a great result.  
![image](https://github.com/user-attachments/assets/8cbba37c-f87f-48f4-8a4d-4ff26f270996)

Nystrata committed:
Update schema.json to specifically mention that it's for Visor HUD elements

Nystrata committed:
Update __init__.py to add comment explaining what the function does

duncathan reviewed (changes requested):
would be great if we can fix the scaling on any text being drawn in the FRMEs (ammo count, energy, etc) 
also ppc-asm has a new release now with the features you need, so please update the requirement

Nystrata committed:
Update pyproject.toml to require ppc-asm 1.5.0 or later

ppc-asm 1.5.0 fixes fmuls and fdivs args

Nystrata committed:
Merge branch 'WidescreenHUDPatch' of https://github.com/Nystrata/open-prime-rando into WidescreenHUDPatch

Nystrata commented:
Hey y'all,
At the moment, the current process for editing FRME is time consuming and I haven't had the time to look into it and relearn the process (it's been a while)
I'm still happy to make patcher changes as reviewed or ship this as is, but I don't think I'll have the time to do FRME changes in the coming weeks.

If y'all are curious, here are my notes for the process I use: https://github.com/Nystrata/EchoesWidescreenHUD/wiki/Notes
